File: attached_assets/libLensAI_PDFGen.py
import argparse
import json
import pandas as pd
from reportlab.lib.pagesizes import A4
from reportlab.pdfgen import canvas
from reportlab.lib.units import cm
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from reportlab.lib.enums import TA_LEFT
from reportlab.lib.colors import black
from io import StringIO
from reportlab.platypus import Frame, Paragraph, KeepInFrame


def load_data(json_file, csv_file):
    data = []
    if json_file:
        with open(json_file, 'r', encoding='utf-8') as jf:
            data.extend(json.load(jf))
    if csv_file:
        csv_data = pd.read_csv(csv_file).to_dict(orient='records')
        data.extend(csv_data)
    return data

# Styles
styles = getSampleStyleSheet()
body_style = ParagraphStyle(
    'body',
    fontName="Times-Roman",
    fontSize=9,
    leading=12,
    alignment=TA_LEFT,
)
summary_style = ParagraphStyle(
    'summary',
    fontName="Times-Roman",
    fontSize=9,
    leading=12,
    alignment=TA_LEFT,
    spaceBefore=6,
    spaceAfter=6,
)

# Fixed height box rendering function
def render_fixed_layout_pdf(data, path):
    c = canvas.Canvas(path, pagesize=A4)
    width, height = A4
    margin = 2 * cm
    gap = 1 * cm
    col_width = (width - 2 * margin - gap) / 2
    columns = [margin, margin + col_width + gap]
    row_height = (height - 2 * margin) / 2
    rows_per_page = 2
    cols_per_page = 2
    items_per_page = rows_per_page * cols_per_page

    for idx, item in enumerate(data):
        page_item = idx % items_per_page
        if page_item == 0 and idx != 0:
            c.showPage()

        row = page_item // 2
        col = page_item % 2

        x = columns[col]
        y = height - margin - row * row_height

        box_height = row_height - 10
        box_width = col_width

        # Draw black border box
        c.setStrokeColor(black)
        c.rect(x, y - box_height, box_width, box_height, stroke=1, fill=0)

        # Text content
        content_y = y - 10
        text_x = x + 5
        max_width = box_width - 10
        text_y = content_y

        paragraphs = [
            f"<b>{item['author']}</b>:",
            f"{item['title']} / {item['author']} ; Illustrationen von {item['illustrator']}.",
            f"- {item['edition']} - {item['publisher']}, {item['year']}.",
            f"{item['pages']} : Illustrationen, farbig ; {item['dimensions']}",
            f"ISBN {item['isbn']} Festeinb. : {item['price']}",
            f"<i>{item['review']}</i>",
            f"{item['reviewer']}",
            f"IK: {item['ik']}",
            f"{item['idb']}",
            f"ASB: {item['asb']}",
            f"<b>Zusammenfassung:</b> {item['summary']}"
        ]

        for text in paragraphs:
            p = Paragraph(text, summary_style if "Zusammenfassung" in text else body_style)
            w, h = p.wrap(max_width, box_height)
            if text_y - h < y - box_height + 5:
                break  # prevent overflow
            p.drawOn(c, text_x, text_y - h)
            text_y -= h + 2

    c.save()

def main():
    parser = argparse.ArgumentParser(description="Generate bibliographic PDF from JSON/CSV data.")
    parser.add_argument("--json", help="Path to JSON input file", default=None)
    parser.add_argument("--csv", help="Path to CSV input file", default=None)
    parser.add_argument("--output", required=True, help="Output PDF file path")
    args = parser.parse_args()

    data = load_data(args.json, args.csv)
    if not data:
        print("No data found in provided input files.")
        return
#    generate_multi_column_pdf(data, args.output)
#    generate_precise_pdf(data, args.output)
#    generate_refined_pdf(data, args.output)
#    render_pdf_with_summaries(data, args.output)
    render_fixed_layout_pdf(data, args.output)


    print(f"PDF generated at: {args.output}")

if __name__ == "__main__":
    main()

File: attached_assets/libLensAI_PDFGen.ts
import { PDFDocument, rgb, StandardFonts } from 'pdfkit';
import { parse } from 'csv-parse/sync';
import * as fs from 'fs';
import * as path from 'path';

interface BookData {
    author: string;
    title: string;
    illustrator: string;
    edition: string;
    publisher: string;
    year: string;
    pages: string;
    dimensions: string;
    isbn: string;
    price: string;
    review: string;
    reviewer: string;
    ik: string;
    idb: string;
    asb: string;
    summary: string;
}

async function loadData(jsonFile?: string, csvFile?: string): Promise<BookData[]> {
    const data: BookData[] = [];
    
    if (jsonFile) {
        const jsonContent = await fs.promises.readFile(jsonFile, 'utf-8');
        data.push(...JSON.parse(jsonContent));
    }
    
    if (csvFile) {
        const csvContent = await fs.promises.readFile(csvFile, 'utf-8');
        const records = parse(csvContent, {
            columns: true,
            skip_empty_lines: true
        });
        data.push(...records);
    }
    
    return data;
}

async function renderFixedLayoutPDF(data: BookData[], outputPath: string): Promise<void> {
    const doc = new PDFDocument({
        size: 'A4',
        margin: 50
    });

    const stream = fs.createWriteStream(outputPath);
    doc.pipe(stream);

    const width = doc.page.width;
    const height = doc.page.height;
    const margin = 50;
    const gap = 30;
    const colWidth = (width - 2 * margin - gap) / 2;
    const columns = [margin, margin + colWidth + gap];
    const rowHeight = (height - 2 * margin) / 2;
    const rowsPerPage = 2;
    const colsPerPage = 2;
    const itemsPerPage = rowsPerPage * colsPerPage;

    for (let idx = 0; idx < data.length; idx++) {
        const pageItem = idx % itemsPerPage;
        if (pageItem === 0 && idx !== 0) {
            doc.addPage();
        }

        const row = Math.floor(pageItem / 2);
        const col = pageItem % 2;

        const x = columns[col];
        const y = height - margin - row * rowHeight;

        const boxHeight = rowHeight - 10;
        const boxWidth = colWidth;

        // Draw border
        doc.rect(x, y - boxHeight, boxWidth, boxHeight)
           .stroke();

        // Text content
        let textY = y - 10;
        const textX = x + 5;
        const maxWidth = boxWidth - 10;

        const paragraphs = [
            { text: `${data[idx].author}:`, bold: true },
            { text: `${data[idx].title} / ${data[idx].author} ; Illustrationen von ${data[idx].illustrator}.` },
            { text: `- ${data[idx].edition} - ${data[idx].publisher}, ${data[idx].year}.` },
            { text: `${data[idx].pages} : Illustrationen, farbig ; ${data[idx].dimensions}` },
            { text: `ISBN ${data[idx].isbn} Festeinb. : ${data[idx].price}` },
            { text: data[idx].review, italic: true },
            { text: data[idx].reviewer },
            { text: `IK: ${data[idx].ik}` },
            { text: data[idx].idb },
            { text: `ASB: ${data[idx].asb}` },
            { text: `Zusammenfassung: ${data[idx].summary}`, bold: true }
        ];

        for (const p of paragraphs) {
            if (textY - 12 < y - boxHeight + 5) break;

            doc.fontSize(9)
               .font(StandardFonts.TimesRoman)
               .text(p.text, textX, textY, {
                   width: maxWidth,
                   align: 'left',
                   continued: false,
                   bold: p.bold,
                   italic: p.italic
               });
            
            textY -= 12;
        }
    }

    doc.end();
}

async function main() {
    const args = process.argv.slice(2);
    const jsonFile = args.find(arg => arg.startsWith('--json='))?.split('=')[1];
    const csvFile = args.find(arg => arg.startsWith('--csv='))?.split('=')[1];
    const outputFile = args.find(arg => arg.startsWith('--output='))?.split('=')[1];

    if (!outputFile) {
        console.error('Output file path is required');
        process.exit(1);
    }

    try {
        const data = await loadData(jsonFile, csvFile);
        if (data.length === 0) {
            console.log('No data found in provided input files.');
            return;
        }

        await renderFixedLayoutPDF(data, outputFile);
        console.log(`PDF generated at: ${outputFile}`);
    } catch (error) {
        console.error('Error generating PDF:', error);
        process.exit(1);
    }
}

if (require.main === module) {
    main();
} 